test/Succeed/UnfoldingWarnings.agda

```diff
@@ -0,0 +1,47 @@
+module UnfoldingWarnings where
+
+open import Agda.Builtin.Nat
+open import Agda.Builtin.Reflection renaming (bindTC to _>>=_)
+open import Agda.Builtin.List
+open import Agda.Builtin.Equality
+
+opaque
+  foo : Nat → Nat
+  foo x = suc x
+
+  bar : Nat → Nat
+  bar x = suc x
+
+opaque unfolding (foo ; bar) where
+  baz : ∀ x → foo x ≡ suc x
+  baz x = refl
+
+opaque unfolding (foo ; bar ; baz) where
+  baz′ : ∀ x → foo x ≡ suc x
+  baz′ x = refl
+
+test : Nat
+test = 0
+
+abstract
+  quux : Nat
+  quux = 0
+
+opaque
+  some-other : Nat
+  some-other = 0
+
+opaque unfolding (quux ; bar ; foo ; baz ; baz′ ; asdf ; ghij ; some-other) where
```

Comment:
The parentheses are superfluent, it seems, or is there an example where they are needed to parse.
Also, `unfolding` should be a layout keyword so that you can drop the semicolons if you like:
```agda
opaque unfolding 
    quux ; bar 
    foo 
    baz ; baz′ 
    asdf ; ghij ; some-other 
  where
...
```

Reply:
I can give this a shot since I have worked with layout before.

Reply:
Is these parentheses supposed to resemble `import using` statements?

Reply:
That was the idea, yes

Reply:
I think the first stab reused the parser for import directives (and thus similar syntax), but in essence, this is a plain identifier list, so no sophisticated parser needed.  
`unfolding` ... `where` is already a pair of brackets, this is why we do not need to nest another pair inside this.

NB. I am not fond of the `using` syntax either, but at least there the parentheses have a function (since there is no closing bracket matching `using`).

Reply:
I'll need a bit time to think how to get this layout situation parsed correctly.  

For now, I have discovered a problem with the current parser:
```agda
opaque unfolding (f)
  where
  foo : ...
  foo = ...
```
This is a parse error; the `where` is not accepted on a new line here.  
However, the user will expect this to work like `module ... where` which can deal with that:
```agda
module M
  where
  postulate A : Set
```

I'll have a look at this problem in connection with the syntax revision.

Reply:
Yea, the parser right now is a bit... explosive. I tried my best but I'm not that familiar with layout. It's made more complicated because `opaque` (unlike `module`) can be a layout keyword by itself.

But that was only implemented because `abstract` already existed and we were adding onto it. If the parser is made overly complicated/fragile by supporting `opaque` without `unfolding`, it might be a good idea to revise the surface syntax? Either way, I appreciate the help 😅

Reply:
One possibility, that AIUI would be easier to parse, was if `opaque` and `unfolding` were separate concrete syntactic structures, each with their own layout contexts:

```agda
opaque -- {
  unfolding -- {
    foo bar 
  -- } closing by the offside rule
  ...
-- } closing by the offside rule
```

Though naïvely that would mean we have to give a semantics for

```agda
opaque
  unfolding foo
  -- some definitions
  unfolding bar
  --- more definitions
```

which is.. it'd have to be done by "lifting" all the unfolding blocks into an unfolding _clause_ to attach to the block, so a bit counter-intuitive, I think.

Reply:
Should the parentheses and semicolons be removed? They are not needed for module declarations:
```agda
module M A B where
  _ : Set
  _ = A × B
```
They are also not needed for data type declarations:
```agda
data D A B C : Set where
  c : A → B → C → D A B C
```
However, they are used in `using` and `hiding` lists, which are rather similar (plain lists of names for which the order does not matter).

I think we should point out clearly that this feature is experimental, and that things can change.

Reply:
> One possibility, that AIUI would be easier to parse, was if `opaque` and `unfolding` were separate concrete syntactic structures, each with their own layout contexts:

A. Yes, I think this one possibility, where `opaque` would be in analogy to e.g. `private`, and `unfolding ... where` would be in analogy to `module ... where`.  You could have then things like:
```agda
opaque
  foo : ...
  foo = ...
  unfolding foo where
    bar : ...
    bar = ...
    baz : ...
    baz = ...
  quux : ...
  quux = ...
```
Syntactically, `unfolding ... where` would be just a `Definition` and a later check would ensure it is only used inside an `opaque` block.

B. The other possibility I see is that we treat `opaque unfolding ... where` as introduction to a block (more in line what exists at the moment).  I think this can be implemented by changing the layout mechanism a bit.  Here, `opaque` should only be a layout keyword _if it is not followed by_ `unfolding`.
Not sure yet how to implement this, whether we can peek at the next token, or whether we have can just pop the layout context introduced by `opaque` when seeing `unfolding` and `opaque` was the previous token.

Reply:
>      A. Yes, I think this one possibility, where opaque would be in analogy to e.g. private, and unfolding ... where would be in analogy to module ... where. You could have then things like:

I think this solution (separating `opaque` and `unfolding ... where`) makes the most sense, it would also allow us to give better error messages.

Reply:
Should `quux` unfold `foo` in the example above? What about `quux` unfolding `bar`, `baz`?

Reply:
I think the default should be that *only* `opaque` definitions that are mentioned explicitly in an enclosing `unfolding ... where` statement should be unfolded (+ transitively the definitions that they unfold). So in the example by @andreasabel above, `quux` should not unfold any `opaque` definitions since it is not inside an `unfolding` block.

Reply:
So:

```agda
opaque
  foo = ... {- unfolds: foo -}
  bar = ... {- unfolds: bar -}

opaque
  unfolding foo where
    baz = ... {- unfolds: foo -}
  unfolding { bar ; baz } where
    quux = ... {- unfolds: bar; foo; baz -}
```

I'm not sure if I like definitions in the _same_ opaque block (i.e. not one in `opaque` and one in a child `unfolding`) not to be able to see each-other.. it'd also require generating `OpaqueId`s for every definition immediately in an opaque block, which is a significant complication over the current implementation (since presumably unfolding `foo` should also unfold anything opaque in `foo`'s where module, if it has one).

Reply:
Sorry, maybe the example I gave was not very good.  I hadn't read the docs.
I infer from the examples in the docs that in one opaque block, later definitions can unfold earlier definitions.  So,
```agda
opaque
  a
  b 
  c
```
is short for 
```agda
opaque a
opaque unfolding a where b
opaque unfolding a; b where c
```
If we keep this semantics, then
```agda
opaque
  a
  unfolding b where c;d
  e
```
would expand to:
```agda
opaque a
opaque unfolding a; b where c
opaque unfolding a; b; c where d
opaque unfolding a; b {-transitively-}; c; d where e
```
Is this roughly what you intended?

Btw, would be great if we could write `unfolding ... in ...`.  However, I don't know if this can be integrated into our layout mechanism, because `in` has a different role there.  Have to have a look.

Reply:
> I'm not sure if I like definitions in the same opaque block (i.e. not one in opaque and one in a child unfolding) not to be able to see each-other..

Ah yes I see the problem. The current problem I had with the prototype `abstract unfolding` from AIM is that all `abstract` definitions in the same *module* can see through each other, which is much too coarse. But perhaps having definitions from the same `opaque` *block* see through each other would not be so bad, and would make the implementation more straightforward as you said.

In that case, `quux` should unfold all of `foo`, `bar`, and `baz` since they are in the same `opaque` block.

Reply:
> Btw, would be great if we could write unfolding ... in .... However, I don't know if this can be integrated into our layout mechanism, because in has a different role there. Have to have a look.

We discussed a local unfolding construct during the AIM, but the consensus was that we should first implement a minimum viable product with just top-level unfoldings, and then look into extending it later.

Reply:
Translating Andreas' example to today's syntax and semantics, we have:
- `foo` and `quux` can unfold each other, 
- `bar` and `baz` can unfold each other, 
- Neither pair of definitions automatically unfolds the other pair.

Nested opaque blocks are treated independently of any enclosing opaque blocks.

Definitions in the same opaque block (in a literal sense) automatically unfold each other, in addition to the transitive closure of the names mentioned in the unfolding clause, if any.

We decided on this because `abstract` does the same, and it's also very smoothly supported by the current implementation.

My question is, having split "opaque block" from "unfolding block" syntactically, should they be semantically split, too? And if so, _how_? Should `unfolding` blocks also unfold everything that belongs to the same enclosing `opaque` block? What about nested `opaque` blocks: should they also unfold lexically-enclosing `opaque` blocks?

Reply:
> We decided on this because `abstract` does the same, and it's also very smoothly supported by the current implementation.

Different top-level `abstract` blocks in the same module are treated as one block. I hope that will not be the case for `opaque` blocks.